### PR TITLE
[ADP- 3166] Add Babbage era specs as fine-types

### PR DESCRIPTION
**Pull request by paolino**
ADP-3166

### CODE REVIEW

File: lib/fine-types/test/Language/FineTypes/Cardano/Ledger/BabbageSpec.hs [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+    moduleMultiFileSpec
+        "Babbage"
+        [ "Crypto.fine"
+        , "PParams.fine"
+        , "Address.fine"
+        , "Block.fine"
+        , "Tx.fine"
+        , "Delegation.fine"
+        , "Script.fine"
+        , "CostModel.fine"
```
Comment (HeinrichApfelmus) [inline lib/fine-types/test/Language/FineTypes/Cardano/Ledger/BabbageSpec.hs:23]:
```suggestion
        , "CostModel.fine"
        , "Value.fine"
```

File: lib/fine-types/test/data/Cardano/Ledger/Babbage/Block.fine [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+    { prev    : HashHeader? --^ hash of previous block header
+    , vk      : VKey --^ block issuer
+    , vrfVk   : VKey  --^ VRF verification key
+    , blockno : BlockNo --^ block number
+    , slot    : Slot --^ block slot
+    , vrfRes  : Seed --^ VRF result value
+    , η       : Seed --^ nonce
+    , prf_η   : Proof --^ nonce proof
+    , ℓ       : {x : ℚ | x ∈ [0, 1]}  --^ leader election value
+    , prf_ℓ   : Proof --^ leader election proof
```
Comment (HeinrichApfelmus) [inline lib/fine-types/test/data/Cardano/Ledger/Babbage/Block.fine:38]:
```suggestion
    , prf     : Proof --^ vrf proof
```

They really removed fields from the Block in Babbage. 🤓 If you ask me, they should have highlighted `prf` in yellow as well. 🤔 Probably difficult to highlight things that have been removed, though.

File: lib/fine-types/test/data/Cardano/Ledger/Babbage/CostModel.fine [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+------------------------------------------------------------------------------}
+
+-- Abstract types
+CostModel   = _ ; --^ Coefficients for the cost model
+
+-- Derived types
+Language    = Σ{PlutusV1 : Unit}; --^ Script Language
+Prices      = ℚ × ℚ; --^ Coefficients for ExUnits prices
+ExUnits     = ℕ × ℕ; --^ Abstract execution units
+LangDepView = Bytes × Bytes; --^ Language Tag and PParams view
```
Comment (HeinrichApfelmus) [inline lib/fine-types/test/data/Cardano/Ledger/Babbage/CostModel.fine:15]:
```suggestion
LangDepView = Bytes × Bytes; --^ Language Tag and PParams view

```

newline 🤫

File: lib/fine-types/test/data/Cardano/Ledger/Babbage/Value.fine [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+    Transactions
+    Mary spec, Figure 3
+------------------------------------------------------------------------------}
+
+AssetName   = Bytes;
+PolicyID    = ScriptHash;
+AdaIDType   = Σ{ AdaID : Unit};
+AssetID     = AdaIDType + (PolicyID × AssetName);
+Quantity    = ℤ;
+Value       = AssetID ↦0 Quantity;
```
Comment (HeinrichApfelmus) [inline lib/fine-types/test/data/Cardano/Ledger/Babbage/Value.fine:15]:
```suggestion
Value       = AssetID ↦0 Quantity;

```

\n 🤫

File: lib/fine-types/test/data/Cardano/Ledger/Babbage/CostModel.fine [hunk 5]
```diff
@@ -0,0 +1,15 @@
+module CostModel where
+
+{-----------------------------------------------------------------------------
+    Protocol parameters
+    Alonzo spec, Figure 1
```
Comment (HeinrichApfelmus) [inline lib/fine-types/test/data/Cardano/Ledger/Babbage/CostModel.fine:5]:
```suggestion
    Alonzo spec, Figure 1
    Babbage spec, Figure 3
```

File: lib/fine-types/test/data/Cardano/Ledger/Babbage/CostModel.fine [hunk 6]
```diff
@@ -0,0 +1,15 @@
...
+------------------------------------------------------------------------------}
+
+-- Abstract types
+CostModel   = _ ; --^ Coefficients for the cost model
+
+-- Derived types
+Language    = Σ{PlutusV1 : Unit}; --^ Script Language
```
Comment (HeinrichApfelmus) [inline lib/fine-types/test/data/Cardano/Ledger/Babbage/CostModel.fine:12]:
```suggestion
Language    =
    Σ{ PlutusV1 : Unit
     , PlutusV2 : Unit
    }; --^ Script Language
```

In the Alonzo spec, they defined `Language = { PlutusV1, … }` in an open world fashion again, so we have to close the world again here. 😅 

In the Babbage spec, Figure 3, the function `allowedLanguages` indicates that the open world includes (and is also restricted to) `PlutusV2`.

File: lib/fine-types/test/data/Cardano/Ledger/Babbage/Tx.fine [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+    { spendInputs           : ℙ TxIn  --^ transaction inputs
+    , collInputs            : ℙ TxIn  --^ collateral inputs
+    , refInputs             : ℙ TxIn  --^ refrerence inputs
+    , txouts                : Ix ↦ TxOut  --^  transaction outputs
+    , collRet               : TxOut?  --^ collateral return output
+    , txcoll                : Coin? --^ total collateral
+    , txcerts               : DCert*  --^ delegation certificates
+    , mint                  : Value  --^ value minted in this transaction
+    , txfee                 : Coin  --^ transaction fee
+    , txvldt                : Slot? × Slot?  --^ validity interval
```
Comment (HeinrichApfelmus) [inline lib/fine-types/test/data/Cardano/Ledger/Babbage/Tx.fine:69]:
```suggestion
    , txvldt                : ValidityInterval  --^ validity interval
```

**Review by HeinrichApfelmus — APPROVED** [review]
Great, thank you! 😊 Several comments to be addressed before merging. Most importantly, they did change the `Block` type, and `PlutusV2` is a new constructor for `Language`.

File: lib/fine-types/test/data/Cardano/Ledger/Babbage/Block.fine (hunk 2 above)
Comment (paolino) [reply lib/fine-types/test/data/Cardano/Ledger/Babbage/Block.fine:38]: Ohh, thanks for really looking into

File: lib/fine-types/test/data/Cardano/Ledger/Babbage/CostModel.fine (hunk 6 above)
Comment (paolino) [reply lib/fine-types/test/data/Cardano/Ledger/Babbage/CostModel.fine]: digging into the function types to guess the existence of type, open up to other potential misses ... Thanks!

**Pull request merged**